providers/identity_provider.go

```diff
@@ -1,13 +1,25 @@
 package providers
 
 import (
+	v1client "github.com/rancher/go-rancher/client"
 	"github.com/rancher/go-rancher/v2"
 	"github.com/rancher/rancher-auth-service/model"
 	"github.com/rancher/rancher-auth-service/providers/github"
 	ad "github.com/rancher/rancher-auth-service/providers/ldap/ad"
 	"github.com/rancher/rancher-auth-service/providers/shibboleth"
 )
 
+//Providers map
+var Providers map[string]IdentityProvider
+
+//RegisterProviders creates object of type driver for every request
+func RegisterProviders() {
+	Providers = map[string]IdentityProvider{}
+	Providers["githubconfig"] = &github.GProvider{}
+	Providers["shibbolethconfig"] = &shibboleth.SProvider{}
+	Providers["ldapconfig"] = &ad.ADProvider{}
+}
+
```

Comment:
Instead of exposing IdentityProvider instances directly in the map here, we should always get an instance of a provider by calling GetProvider(name string).

So you can instead have a list of providerNames stored in an array in place of Providers map[string]IdentityProvider.
In auth_server you can add a method GetProvider(providerName) that in turn invokes
newProvider := providers.GetProvider(providerName) and return a provider instance to your getSchemas() method below.
Use that instance to invoke GetProviderConfigResource() and CustomizeSchema() methods. This way we don't instantiate providers and save them in the map for lifetime of the auth_service.